File: go/internal/twiml/verbs.go
// Package twiml implements a strict, hand-rolled parser and verb dispatcher for
// the Twilio Markup Language (TwiML) subset that v3.0 supports: <Response> root
// containing <Connect><Stream>, <Dial>+<Number>, <Hangup>, <Redirect>, <Reject>.
//
// The parser uses a token-walk over encoding/xml.Decoder.Token() rather than
// struct tags with the wildcard child marker. Reasons (per
// a known XML-pitfall): Go's encoding/xml populates BOTH
// the named-field pointer AND the wildcard slice for the same element,
// double-counting verbs. The token walker gives perfect control over which
// verbs are accepted and the order they appear in the dispatch list.
//
// Unknown verbs are retained as `unknownVerb` wrappers so the dispatcher can
// emit a per-verb warn-and-skip log (TWIML-05).
package twiml

import (
	"encoding/xml"
	"strings"
)

// Verb is a TwiML verb produced by the parser. Each concrete verb type below
// (Connect, Dial, Hangup, Redirect, Reject) implements XMLName() returning a
// stable element-local name used by the dispatcher to look up handlers.
type Verb interface {
	XMLName() xml.Name
}

// Response is the parsed root document. Verbs is in document order — the
// dispatcher walks the slice in order.
type Response struct {
	Verbs []Verb
}

// Connect carries a single nested <Stream>. v3.0 only supports <Stream> as a
// Connect child; other Connect children would be parsed as unknown and
// warned-and-skipped by the dispatcher.
type Connect struct {
	Stream *Stream
}

// XMLName implements Verb. Local must match the dispatcher handler key.
func (Connect) XMLName() xml.Name { return xml.Name{Local: "Connect"} }

// Stream models <Stream url="…" name="…" track="…"> with optional
// <Parameter name="…" value="…"/> children flattened into Parameters.
type Stream struct {
	URL        string
	Name       string
	Track      string
	Parameters map[string]string
}

// Dial supports both bare-chardata form (<Dial>+49…</Dial>) AND a <Number>
// child form (<Dial><Number>+49…</Number></Dial>) — see ResolveDialTarget for
// the disambiguation rules. The dispatcher routes Dial to the B2BUA
// forwarder (or to notImplHandler when forwarding is not wired).
type Dial struct {
	CallerID       string
	Timeout        *int
	TimeLimit      *int
	HangupOnStar   bool
	Action         string
	Method         string
	AnswerOnBridge *bool

	// Mutually-preferable target sources — at most one populated in a valid doc.
	NumberText string  // <Dial>+49…</Dial>
	Number     *Number // <Dial><Number>+49…</Number></Dial>

	// Rejected child markers — parser detects these for the dispatcher to warn on.
	HasSip        bool
	HasClient     bool
	HasConference bool
	HasQueue      bool

	// Parent-<Dial> status-callback subscription. Per-<Number> overrides
	// take precedence at DialOpts construction time (verb_dial.go).
	StatusCallback       string   // <Dial statusCallback="…">
	StatusCallbackMethod string   // "POST" | "GET"; empty defaults to POST at emission
	StatusCallbackEvents []string // tokenized event-name list; nil/empty = subscribe-to-all
}

// XMLName implements Verb.
func (Dial) XMLName() xml.Name { return xml.Name{Local: "Dial"} }

// Number is the <Number> child of <Dial>. Per-leg status-callback
// subscription attributes win over the parent <Dial> values.
// Backward compat: a bare-text <Number>+49…</Number> still parses with
// Text populated and the status fields zero-valued.
type Number struct {
	Text string

	// Per-<Number>-leg status-callback overrides. Resolution rule
	// (verb_dial.go): if StatusCallback != "" use it; otherwise fall
	// back to the parent <Dial> value.
	StatusCallback       string   // <Number statusCallback="…">
	StatusCallbackMethod string   // "POST" | "GET"
	StatusCallbackEvents []string // tokenized event-name list
}

// Hangup models <Hangup/>. The dispatcher routes this to hangupHandler which
// invokes DispatcherSession.Hangup() and returns ActionStop.
type Hangup struct{}

// XMLName implements Verb.
func (Hangup) XMLName() xml.Name { return xml.Name{Local: "Hangup"} }

// Redirect models <Redirect method="POST">https://…</Redirect>. Currently
// routed to the notImplHandler stub.
type Redirect struct {
	Method string
	URL    string
}

// XMLName implements Verb.
func (Redirect) XMLName() xml.Name { return xml.Name{Local: "Redirect"} }

// Reject models <Reject reason="busy|rejected"/>. Currently a stub —
// SIP-final-response mapping is not yet wired.
type Reject struct {
	Reason string
}

// XMLName implements Verb.
func (Reject) XMLName() xml.Name { return xml.Name{Local: "Reject"} }

// unknownVerb is the parser's wrapper for any element name inside <Response>
// that does not match a known verb. Retained in Response.Verbs so the
// dispatcher's warn-and-skip path sees them.
type unknownVerb struct {
	Name xml.Name
}

// XMLName implements Verb.
func (u unknownVerb) XMLName() xml.Name { return u.Name }

// ResolveDialTarget returns the dialed number from a Dial verb.
// Bare-text and <Number> child are equivalent; if BOTH are populated the
// caller is logged "ambiguous=true" and the <Number> child wins (matches
// Twilio behavior where structured data trumps inline chardata).
func (d *Dial) ResolveDialTarget() (target string, ambiguous bool) {
	text := strings.TrimSpace(d.NumberText)
	switch {
	case d.Number != nil && text != "":
		return strings.TrimSpace(d.Number.Text), true
	case d.Number != nil:
		return strings.TrimSpace(d.Number.Text), false
	case text != "":
		return text, false
	default:
		return "", false
	}
}

File: node/src/twiml/verbs.ts
/**
 * TwiML verb model — pure data types produced by the parser (parse.ts) and
 * consumed by the dispatcher (dispatch.ts).
 *
 * Port of go/internal/twiml/verbs.go. The verb set mirrors the v3 Go parser:
 * <Response> root containing <Connect><Stream>, <Dial>+<Number>, <Hangup>,
 * <Redirect>, <Reject>. Unknown verbs are retained as UnknownVerb so the
 * dispatcher can warn-and-skip them (never silently dropped).
 */

/** Discriminant tag for each verb kind. Mirrors the XML element local-name. */
export type VerbKind =
  | 'Connect'
  | 'Dial'
  | 'Hangup'
  | 'Redirect'
  | 'Reject'
  | 'Unknown';

/** Common shape: every verb carries its element-local name for dispatch logs. */
interface VerbBase {
  readonly kind: VerbKind;
  /** Element-local name (e.g. "Dial", "Say"). Used by the dispatcher warn logs. */
  readonly name: string;
}

/**
 * Number is the <Number> child of <Dial>. Per-leg status-callback subscription
 * attributes win over the parent <Dial> values (resolved in verbDial.ts).
 * Backward compat: a bare-text <Number>+49…</Number> parses with text set and
 * the status fields undefined.
 */
export interface Number {
  text: string;
  statusCallback?: string;
  statusCallbackMethod?: string;
  statusCallbackEvents?: string[];
}

/**
 * Dial supports both bare-chardata form (<Dial>+49…</Dial>) AND a <Number>
 * child form (<Dial><Number>+49…</Number></Dial>) — see resolveDialTarget for
 * the disambiguation rules.
 */
export interface Dial extends VerbBase {
  readonly kind: 'Dial';
  readonly name: 'Dial';
  callerId?: string;
  /** Ring timeout in seconds; undefined → default applies (30s). */
  timeout?: number;
  /** Max talk time in seconds; undefined → default applies (14400s). */
  timeLimit?: number;
  hangupOnStar: boolean;
  action?: string;
  method?: string;
  answerOnBridge?: boolean;

  // Status-callback subscription on the parent <Dial>. Per-<Number> overrides
  // take precedence at DialOpts construction time (verbDial.ts).
  statusCallback?: string;
  statusCallbackMethod?: string;
  statusCallbackEvents?: string[];

  // Mutually-preferable target sources — at most one populated in a valid doc.
  numberText?: string;
  number?: Number;

  // Rejected child markers — parser detects these for the dispatcher to warn on.
  hasSip: boolean;
  hasClient: boolean;
  hasConference: boolean;
  hasQueue: boolean;
}

/** Stream models <Stream url=… name=… track=…> with flattened <Parameter> children. */
export interface Stream {
  url?: string;
  name?: string;
  track?: string;
  parameters?: Record<string, string>;
}

/** Connect carries a single nested <Stream>. */
export interface Connect extends VerbBase {
  readonly kind: 'Connect';
  readonly name: 'Connect';
  stream?: Stream;
}

/** Hangup models <Hangup/>. Terminal verb. */
export interface Hangup extends VerbBase {
  readonly kind: 'Hangup';
  readonly name: 'Hangup';
}

/** Redirect models <Redirect method="POST">https://…</Redirect>. */
export interface Redirect extends VerbBase {
  readonly kind: 'Redirect';
  readonly name: 'Redirect';
  method?: string;
  url?: string;
}

/** Reject models <Reject reason="busy|rejected"/>. */
export interface Reject extends VerbBase {
  readonly kind: 'Reject';
  readonly name: 'Reject';
  reason?: string;
}

/**
 * UnknownVerb wraps any element inside <Response> that is not a known verb.
 * Retained so the dispatcher's warn-and-skip path sees it (never dropped).
 */
export interface UnknownVerb extends VerbBase {
  readonly kind: 'Unknown';
}

/** Verb is the discriminated union of all parsed verbs. */
export type Verb = Connect | Dial | Hangup | Redirect | Reject | UnknownVerb;

/**
 * Response is the parsed root document. verbs is in document order — the
 * dispatcher walks the array in order.
 */
export interface Response {
  verbs: Verb[];
}

/**
 * resolveDialTarget returns the dialed number from a Dial verb.
 * Bare-text and <Number> child are equivalent; if BOTH are populated the
 * caller is logged "ambiguous=true" and the <Number> child wins (matches
 * Twilio behavior where structured data trumps inline chardata). Whitespace
 * is trimmed.
 */
export function resolveDialTarget(d: Dial): { target: string; ambiguous: boolean } {
  const text = (d.numberText ?? '').trim();
  if (d.number !== undefined && text !== '') {
    return { target: d.number.text.trim(), ambiguous: true };
  }
  if (d.number !== undefined) {
    return { target: d.number.text.trim(), ambiguous: false };
  }
  if (text !== '') {
    return { target: text, ambiguous: false };
  }
  return { target: '', ambiguous: false };
}
